From a0bc83baad16f0e5326cf8c428f102954331500b Mon Sep 17 00:00:00 2001
From: Alex Bee <knaerzche@gmail.com>
Date: Mon, 4 Jan 2021 22:38:26 +0100
Subject: [PATCH 56/92] drm/rockchip: seperate mode clock validation

seperate mode clock validation between internal and external
phy types.
this will allow modes >= 2160p@50Hz on RK3288/RK3399 (RGB444)

Signed-off-by: Alex Bee <knaerzche@gmail.com>
---
 drivers/gpu/drm/rockchip/dw_hdmi-rockchip.c | 18 ++++++++++++++++--
 1 file changed, 16 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/rockchip/dw_hdmi-rockchip.c b/drivers/gpu/drm/rockchip/dw_hdmi-rockchip.c
index 42457f7d9bc9..90cc3b33e2a0 100644
--- a/drivers/gpu/drm/rockchip/dw_hdmi-rockchip.c
+++ b/drivers/gpu/drm/rockchip/dw_hdmi-rockchip.c
@@ -325,16 +325,30 @@ dw_hdmi_rockchip_mode_valid(struct dw_hdmi *hdmi, void *data,
 			    const struct drm_display_mode *mode)
 {
 	struct dw_hdmi_plat_data *pdata = (struct dw_hdmi_plat_data *)data;
+	const struct dw_hdmi_mpll_config *mpll_cfg = pdata->mpll_cfg;
+
 	int clock = mode->clock;
+	int i = 0;
 
 	if (pdata->ycbcr_420_allowed && drm_mode_is_420(info, mode) &&
-	    (info->color_formats & DRM_COLOR_FORMAT_YCBCR420))
+	    (info->color_formats & DRM_COLOR_FORMAT_YCBCR420)) {
 		clock /= 2;
+		mpll_cfg = pdata->mpll_cfg_420;
+	}
 
-	if (clock > 340000 ||
+	if ((!mpll_cfg && clock > 340000) ||
 	    (info->max_tmds_clock && clock > info->max_tmds_clock))
 		return MODE_CLOCK_HIGH;
 
+	if (mpll_cfg) {
+		while ((clock * 1000) < mpll_cfg[i].mpixelclock &&
+		       mpll_cfg[i].mpixelclock != (~0UL))
+		       i++;
+
+		if (mpll_cfg[i].mpixelclock == (~0UL))
+			return MODE_CLOCK_HIGH;
+	}
+
 	return drm_mode_validate_size(mode, 3840, 2160);
 }
 static void
-- 
2.34.1

